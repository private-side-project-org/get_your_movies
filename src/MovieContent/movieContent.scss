@use "assets/variables" as *;

.moviesContent {
  // entire screen
  &-container {
    display: flex;
    height: 100vh;
    @include mq(md) {
      grid-template-columns: 40% 60%;
    }
  }

  // no result(both search and favorite)
  &-no-result {
    display: flex;
    flex-direction: column;
    align-items: center;
    text-align: center;

    > img {
      width: 60px;
      margin-bottom: 16px;
    }
  }

  // list tab
  &-tabs {
    display: flex;
    gap: 16px;
    margin-right: 8px;

    > h4 {
      border: 1px solid lightgray;
      padding: 4px 8px;
      border-radius: 6px 6px 0px 0px;
      border-bottom: none;
      cursor: pointer;

      &:hover {
        opacity: 0.6;
      }
    }

    .active {
      background-color: lightblue;
    }
  }

  // left panel container
  &-left-panel-container {
    display: flex;
    flex-direction: column;
    position: relative;
    height: calc(100vh - 20px);
    padding: 16px 0px 0px 8px;
    overflow: scroll;
    flex: 1;

    @include mq(md) {
      flex: 1;
      max-width: 510px;
    }

    // not display below 768px
    &.movie-selected {
      flex: 0;
      @include mq(md) {
        flex: 1;
      }
    }

    // search bar
    > input {
      line-height: 16px;
      font-size: 16px;
      padding: 8px 16px;
      border-radius: 6px;
      border: 2px solid gray;
      margin-bottom: 16px;
      margin-right: 8px;
    }
  }

  // list wrapper
  &-list-wrapper {
    border: 1px solid lightgray;
    flex: 1;
    border-radius: 0px 6px 6px 6px;
    padding: 8px 8px 0px 8px;
    overflow: scroll;
    scroll-behavior: smooth;
    margin-right: 8px;

    &.no-result {
      display: flex;
      justify-content: center;
      align-items: center;
    }
  }

  // anchor
  &-anchor-to-top {
    display: flex;
    align-items: center;
    justify-content: center;
    position: fixed;
    z-index: 10;
    right: 24px;
    bottom: 24px;
    width: 60px;
    height: 60px;
    border: 2px solid lightgray;
    border-radius: 50px;
    background-color: white;

    > span {
      margin: auto;
    }
  }
}
