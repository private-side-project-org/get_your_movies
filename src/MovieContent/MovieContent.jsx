import React, { useState } from "react";
import useSearchMovies from "../queries/useSearchMovies";
import MovieList from "./MovieList/MovieList";
import MoviePanel from "./MoviePanel/MoviePanel";
import debounce from "lodash/debounce";
import CONSTANTS from "utils/constants";
import useSavedMovies from "hooks/useSavedMovies";

import "./movieContent.scss";

const clapperBoard = require("assets/icons/clapperboard.svg");
const { FAVORITE, SEARCH } = CONSTANTS.TAB_OPTIONS;

const MovieContent = () => {
  const [search, setSearch] = useState("");
  const [selectedMovie, setSelectedMovie] = useState(null);
  const [selectedTab, setSelectedTab] = useState(SEARCH);

  // query to get search result
  const {
    searchedMovies,
    loadingMovies,
    hasNextPage,
    fetchNextPage,
    isFetchingNextPage,
  } = useSearchMovies({ search });

  // context movie list from localstorage
  const { favoriteMovieList } = useSavedMovies();

  const handleSearch = debounce((e) => {
    setSearch(e.target.value);
  }, 500);

  const handleClickTab = (tab) => {
    if (tab !== selectedTab) {
      setSelectedTab(tab);
    }
  };

  const movielist =
    selectedTab === FAVORITE && favoriteMovieList
      ? favoriteMovieList
      : searchedMovies;

  // list field render
  const movieListRender = () => {
    // render no favorite
    if (selectedTab === FAVORITE && movielist.length === 0) {
      return (
        <div className="moviesContent-no-result">
          <img src={clapperBoard} alt="clapper_board" />
          <h2>No favorite movie yet</h2>
          <p>You can save your favorite movies here.</p>
        </div>
      );
    }

    return (
      <>
        {!loadingMovies && movielist?.length > 0 ? (
          // render search result
          <MovieList
            movies={movielist}
            hasNextPage={hasNextPage}
            fetchNextPage={fetchNextPage}
            isFetchingNextPage={isFetchingNextPage}
            onSetSelectedMovie={setSelectedMovie}
            selectedTab={selectedTab}
          />
        ) : (
          // render no result found
          <div className="moviesContent-no-result">
            <img src={clapperBoard} alt="clapper_board" />
            <h2>No result found</h2>
            <p>
              Or, you didn&apos;t try search? <br /> Get your movie from search
              bar.
            </p>
          </div>
        )}
      </>
    );
  };

  return (
    <div className="moviesContent-container">
      <div
        className={`moviesContent-left-panel ${
          selectedMovie ? "movie-selected" : ""
        }`}
      >
        <h2>Movie Search</h2>
        <input onChange={handleSearch} placeholder="type keyword here..." />
        <div className="moviesContent-tabs">
          <h4
            className={selectedTab === SEARCH ? "active" : ""}
            onClick={() => handleClickTab(SEARCH)}
          >
            Search result
          </h4>
          <h4
            className={`${selectedTab === FAVORITE ? "active" : ""}`}
            onClick={() => handleClickTab(FAVORITE)}
          >
            Your favorites
          </h4>
        </div>
        <div
          className={`moviesContent-list-wrapper ${
            movielist.length === 0 ? "no-result" : ""
          }`}
        >
          {movieListRender()}
        </div>
        <a href="#top" className="moviesContent-anchor-to-top">
          <span>Top</span>
        </a>
      </div>
      {selectedMovie && (
        <MoviePanel
          selectedMovie={selectedMovie}
          onSetSelectedMovie={setSelectedMovie}
        />
      )}
    </div>
  );
};

export default MovieContent;
