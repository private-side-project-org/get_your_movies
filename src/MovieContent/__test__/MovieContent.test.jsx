import React from "react";
import PropTypes from "prop-types";
import { fireEvent, render, screen } from "@testing-library/react";
import MovieContent from "../MovieContent";
import { QueryClientProvider, QueryClient } from "react-query";
import _ from "lodash";

_.debounce = jest.fn();

jest.mock("../../queries/useSearchMovies", () => {
  return jest.fn(() => ({
    searchedMovies: [],
    loadingMovies: false,
    fetchNextPage: jest.fn(),
    hasNextPage: false,
    isFetchingNextPage: false,
  }));
});

const client = new QueryClient();

const WrapperComponent = ({ children }) => (
  <QueryClientProvider client={client}>{children}</QueryClientProvider>
);

WrapperComponent.propTypes = {
  children: PropTypes.node.isRequired,
};

describe("MovieContent test", () => {
  it("snapshots movieContent", async () => {
    const renderedComponent = render(<MovieContent />, {
      wrapper: WrapperComponent,
    });
    expect(renderedComponent.baseElement).toMatchSnapshot();
  });

  it("set value in input", async () => {
    render(<MovieContent />, {
      wrapper: WrapperComponent,
    });

    const searchInput = await screen.findByTestId("search_input");
    fireEvent.change(searchInput, { target: { value: "abc" } });
    expect(searchInput.value).toBe("abc");
  });

  it("switch favorite tab", async () => {
    render(<MovieContent />, {
      wrapper: WrapperComponent,
    });

    const favoriteTab = await screen.findByTestId("favorite_tab");
    fireEvent.click(favoriteTab);
    expect(favoriteTab.className).toBe("active");
  });

  it("switch search tab", async () => {
    render(<MovieContent />, {
      wrapper: WrapperComponent,
    });

    const searchTab = await screen.findByTestId("search_tab");
    fireEvent.click(searchTab);
    expect(searchTab.className).toBe("active");
  });
});
