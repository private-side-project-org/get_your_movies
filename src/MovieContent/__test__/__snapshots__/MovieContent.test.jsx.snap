// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieContent test snapshots movieContent 1`] = `
<body>
  <div>
    <div
      class="moviesContent-container"
    >
      <div
        class="moviesContent-left-panel-container "
      >
        <h2>
          Movie Search
        </h2>
        <input
          data-testid="search_input"
          placeholder="type keyword here..."
        />
        <div
          class="moviesContent-tabs"
        >
          <h4
            class="active"
            data-testid="search_tab"
          >
            Search result
          </h4>
          <h4
            class=""
            data-testid="favorite_tab"
          >
            Your favorites
          </h4>
        </div>
        <div
          class="moviesContent-list-wrapper no-result"
        >
          <div
            class="moviesContent-no-result"
          >
            <img
              alt="clapper_board"
              src="[object Object]"
            />
            <h2>
              No result found
            </h2>
            <p>
              Or, you didn't try search? 
              <br />
               Get your movie from search bar.
            </p>
          </div>
        </div>
        <a
          class="moviesContent-anchor-to-top"
          href="#top"
        >
          <span>
            Top
          </span>
        </a>
      </div>
    </div>
  </div>
</body>
`;
